language/move-vm/runtime/src/session.rs

```diff
@@ -110,6 +110,29 @@ impl<'r, 'l, R: RemoteCache> Session<'r, 'l, R> {
         self.runtime.publish_module(
             module,
             sender,
+            false,
+            false,
+            &mut self.data_cache,
+            cost_strategy,
+            log_context,
+        )
+    }
+
+    // TODO: merge with publish_module?
+    pub fn publish_module_ext(
+        &mut self,
+        module: Vec<u8>,
+        sender: AccountAddress,
+        allow_republish: bool,
+        allow_breaking_changes: bool,
```

Comment:
I feel uncomfortable with having this flag in the public API even if we plan to remove it later. Thoughts on eliminating it and supporting it via CLI changes earlier instead of later?

Similar with `allow_republish`-- I think it's reasonable to assume that all modules can be republished for now, then add restrictions on republishing inside the VM if/when we change the bytecode format. In general, I don't think we should extend the VM's public API with functionality that we plan to remove later.

Reply:
Totally agreed, in fact, changes on the way!

Reply:
Done, the recent commit should address the discomfort :)

Reply:
> I think it's reasonable to assume that all modules can be republished for now
I don't think we should do this. Republishing a module is always unsafe as it can change the contract in ways we couldn't detect, even if the old and new signatures match. We must make sure that one can not update a module by simply submitting another Diem module publishing transaction. I feel like at least in the short term, the power to republish will need to kept in check.

Reply:
I agree with this in the long term. But I think we do not have any mechanism for enforcing this until we add an `updgradability` byte to the module format. Absent that, we must either assume that every module is updatable or no module is updatable. It's going to be quite hard to use/test the update feature if we assume the latter :).

Reply:
Ah I see. The upgradability byte makes sense to me, especially in an open publishing setup. I would also imagine having more fine-grained control over individual functions or structs. Definitely a feature we should look more into next!

> It's going to be quite hard to use/test the update feature if we assume the latter :).
I still don't quite get this part. The current version of this PR will change the way the existing module publishing transactions work, and allowing contracts to be modified unilaterally. I'm totally fine adding the ability to republish a module, but want to see it being isolated from the normal flow and require special privilege on the Diem layer.

> Absent that, we must either assume that every module is updatable or no module is updatable.
The way I'm understanding this is that in the short term, there will be no open publishing, and since there are a limited number of modules and participants in the network, the association will be able to make folks agree on an upgrade or even change of semantics in certain cases and apply the changes. In other words, all modules are updatable in a while. However since such updates are still extremely dangerous, I just feel like we should use a separate transaction type for it.

Reply:
> We must make sure that one can not update a module by simply submitting another Diem module publishing transaction.

> allowing contracts to be modified unilaterally. I'm totally fine adding the ability to republish a module, but want to see it being isolated from the normal flow

What kind of isolation do you have in mind?

>  and require special privilege on the Diem layer.
> However since such updates are still extremely dangerous, I just feel like we should use a separate transaction type for it.

In the Diem layer, module publishing transactions are already restricted to 0xA550C18 (which is the strictest permission we have). With this PR, 0xA550C18 will also have the ability to update modules. I don't think a new transaction type adds any extra safety on top of this already-strictest level.

Reply:
> The current version of this PR will change the way the existing module publishing transactions work

I am not sure that I get this fully. If the meaning is now publishing and re-publishing follow the same VM API, yes, it is. But I don't think other parts are affected. In particular, both with and without this PR, the transaction for publishing a module will lead to a `TransactionEffect` containing serialized module bytes and the state/storage layer can pick up the bytes and save them accordingly. This part of the module publishing flow is not affected.

Reply:
> the transaction for publishing a module will lead to a TransactionEffect containing serialized module byte
It's when a module already exists things are different:
- Old behavior: the VM will return an error and won't generate a TransactioEffect a all
- New behavior: the VM will generate a TransactionEffects including an updated version of the module

But as Sam mentioned, the Move VM APIs sit behind the Diem ones, so it's not like this can be invoked arbitrarily by a client -- you need association approval to publish anything right now.